In [18]:
import numpy as np
import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler, LabelEncoder
import seaborn as sns

In [2]:
train_df = pd.read_csv(Path('Resources/2019loans.csv'))
test_df = pd.read_csv(Path('Resources/2020Q1loans.csv'))
train_df.head(10)

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
5       31295   31295     6500.0    0.1614       228.98       MORTGAGE   
6      173238  173238    40000.0    0.1240       897.89       MORTGAGE   
7      116755  116755     5000.0    0.1719       178.74       MORTGAGE   
8      170311  170311     6000.0    0.1102       196.49            OWN   
9      254794  254794    10625.0    0.1102       347.95           RENT   

   annual_inc verification_status loan_status pymnt_plan  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified    low_risk          n  ...           100.0   
1    123000.0     Source Verified    low_risk          n  ...            85.0   
2    197000.0     Source Verified    low_risk          n  ...            85.7   
3     45000.0        Not Verified    low_risk          n  ...           100.0   
4    133000.0     Source Verified    low_risk          n  ...           100.0   
5     65000.0        Not Verified    low_risk          n  ...            92.3   
6    171000.0        Not Verified    low_risk          n  ...            88.9   
7     55000.0        Not Verified    low_risk          n  ...            90.0   
8     36000.0        Not Verified    low_risk          n  ...           100.0   
9     73000.0     Source Verified    low_risk          n  ...            91.9   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   
5              50.0                   1.0        0.0          64660.0   
6              75.0                   0.0        0.0         772030.0   
7               0.0                   0.0        0.0          17075.0   
8               0.0                   1.0        0.0          65200.0   
9              30.0                   0.0        0.0          41869.0   

   total_bal_ex_mort  total_bc_limit total_il_high_credit_limit  \
0           122018.0         32000.0                   170200.0   
1            27896.0         15900.0                    35398.0   
2           114043.0         22600.0                    90340.0   
3            20761.0         19900.0                    15406.0   
4           109056.0         79500.0                    58778.0   
5            39487.0         12800.0                    42060.0   
6            83509.0         41200.0                    81300.0   
7            11244.0           500.0                    14275.0   
8            21103.0         43300.0                    12100.0   
9            11859.0         24500.0                    10169.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  
5              N                     N  
6              N                     N  
7              N                     N  
8              N                     N  
9              N                     N  

[10 rows x 86 columns]

My prediction for the more accurate model, when comparing Logistical Regression and Random Forest, is RF because it has many more factors and variables which allow RF to be more maleable and react to the model rather than having forced linear elements which can skew the model. RF can also make decisions in it's Decision Tree to react, while LR does not have the capability.

In [7]:
# Convert categorical data to numeric and separate loan_status feature for training data
y_train=train_df["loan_status"]
X_train= train_df.drop(columns=["loan_status"])
X_train.head(10)

Unnamed: 0   index  loan_amnt  int_rate  installment home_ownership  \
0       57107   57107    13375.0    0.1797       483.34       MORTGAGE   
1      141451  141451    21000.0    0.1308       478.68       MORTGAGE   
2      321143  321143    20000.0    0.1240       448.95       MORTGAGE   
3       11778   11778     3000.0    0.1240       100.22           RENT   
4      169382  169382    30000.0    0.1612      1056.49       MORTGAGE   
5       31295   31295     6500.0    0.1614       228.98       MORTGAGE   
6      173238  173238    40000.0    0.1240       897.89       MORTGAGE   
7      116755  116755     5000.0    0.1719       178.74       MORTGAGE   
8      170311  170311     6000.0    0.1102       196.49            OWN   
9      254794  254794    10625.0    0.1102       347.95           RENT   

   annual_inc verification_status pymnt_plan    dti  ...  pct_tl_nvr_dlq  \
0    223000.0        Not Verified          n  29.99  ...           100.0   
1    123000.0     Source Verified          n  11.26  ...            85.0   
2    197000.0     Source Verified          n  11.28  ...            85.7   
3     45000.0        Not Verified          n  18.08  ...           100.0   
4    133000.0     Source Verified          n  27.77  ...           100.0   
5     65000.0        Not Verified          n  27.49  ...            92.3   
6    171000.0        Not Verified          n  20.74  ...            88.9   
7     55000.0        Not Verified          n  10.80  ...            90.0   
8     36000.0        Not Verified          n  27.28  ...           100.0   
9     73000.0     Source Verified          n   7.09  ...            91.9   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0              50.0                   0.0        0.0         577150.0   
1              33.3                   0.0        0.0         132750.0   
2              33.3                   0.0        0.0         628160.0   
3              16.7                   1.0        0.0          42006.0   
4              66.7                   0.0        0.0         283248.0   
5              50.0                   1.0        0.0          64660.0   
6              75.0                   0.0        0.0         772030.0   
7               0.0                   0.0        0.0          17075.0   
8               0.0                   1.0        0.0          65200.0   
9              30.0                   0.0        0.0          41869.0   

   total_bal_ex_mort total_bc_limit  total_il_high_credit_limit  \
0           122018.0        32000.0                    170200.0   
1            27896.0        15900.0                     35398.0   
2           114043.0        22600.0                     90340.0   
3            20761.0        19900.0                     15406.0   
4           109056.0        79500.0                     58778.0   
5            39487.0        12800.0                     42060.0   
6            83509.0        41200.0                     81300.0   
7            11244.0          500.0                     14275.0   
8            21103.0        43300.0                     12100.0   
9            11859.0        24500.0                     10169.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  
5              N                     N  
6              N                     N  
7              N                     N  
8              N                     N  
9              N                     N  

[10 rows x 85 columns]

In [9]:
# Convert categorical data to numeric and separate loan_status feature for testing data
y_test=test_df["loan_status"]
X_test= test_df.drop(columns=["loan_status"])
X_test.head(10)

Unnamed: 0  index  loan_amnt  int_rate  installment home_ownership  \
0       67991  67991    40000.0    0.0819       814.70       MORTGAGE   
1       25429  25429     6000.0    0.1524       208.70           RENT   
2       38496  38496     3600.0    0.1695       128.27           RENT   
3       19667  19667    20000.0    0.1524       478.33           RENT   
4       37505  37505     3600.0    0.1240       120.27           RENT   
5       26517  26517    27000.0    0.0819       549.93           RENT   
6       30063  30063    15000.0    0.0881       310.00       MORTGAGE   
7       31659  31659    19000.0    0.0881       602.52       MORTGAGE   
8       15003  15003    27575.0    0.1171       912.07            OWN   
9       75464  75464    31500.0    0.1774       795.45       MORTGAGE   

   annual_inc verification_status pymnt_plan    dti  ...  pct_tl_nvr_dlq  \
0    140000.0        Not Verified          n  19.75  ...            97.7   
1     55000.0        Not Verified          n  11.52  ...            66.7   
2     42000.0        Not Verified          n   6.74  ...           100.0   
3    100000.0        Not Verified          n  12.13  ...           100.0   
4     50000.0        Not Verified          n  16.08  ...           100.0   
5     57000.0            Verified          n  31.41  ...            96.2   
6     50000.0        Not Verified          n  24.44  ...           100.0   
7     70000.0        Not Verified          n  30.94  ...           100.0   
8    142000.0     Source Verified          n  17.90  ...           100.0   
9    100000.0     Source Verified          n  25.05  ...            96.2   

   percent_bc_gt_75  pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  \
0               0.0                   0.0        0.0         527975.0   
1               0.0                   0.0        0.0          34628.0   
2               0.0                   0.0        0.0          23100.0   
3              50.0                   0.0        0.0          56481.0   
4              25.0                   0.0        0.0          45977.0   
5              14.3                   0.0        0.0         125240.0   
6               0.0                   0.0        0.0         192786.0   
7              33.3                   0.0        0.0         782289.0   
8              22.2                   1.0        0.0         488561.0   
9              28.6                   0.0        0.0         562627.0   

   total_bal_ex_mort total_bc_limit  total_il_high_credit_limit  \
0            70914.0        74600.0                     99475.0   
1            23460.0         5900.0                     23628.0   
2            19183.0         7300.0                     15000.0   
3            43817.0        13800.0                     35981.0   
4            32448.0        21000.0                     24977.0   
5            52062.0        98400.0                     11629.0   
6            49901.0        43600.0                     43276.0   
7           161240.0        44000.0                    105289.0   
8            89586.0        89500.0                     72659.0   
9           108891.0        27900.0                    115627.0   

   hardship_flag  debt_settlement_flag  
0              N                     N  
1              N                     N  
2              N                     N  
3              N                     N  
4              N                     N  
5              N                     N  
6              N                     N  
7              N                     N  
8              N                     N  
9              N                     N  

[10 rows x 85 columns]

In [12]:
# add missing dummy variables to testing set
X_dummies_train = pd.get_dummies(X_train)
X_dummies_test = pd.get_dummies(X_test)
print(f"Train: {X_dummies_train.shape}, Test: {X_dummies_test.shape}")

Train: (12180, 94), Test: (4702, 93)


In [28]:
y_label_train = LabelEncoder().fit_transform(train_df['loan_status'])
y_label_train

y_label_test = LabelEncoder().fit_transform(test_df['loan_status'])
y_label_test

array([1, 1, 1, ..., 0, 0, 0])

In [29]:
for col in X_dummies_train.columns:
    if col not in X_dummies_test.columns:
        X_dummies_test[col]=0

In [37]:
# Train the Logistic Regression model on the unscaled data and print the model score
from sklearn.linear_model import LogisticRegression
classifier_logistic  = LogisticRegression(solver='lbfgs').fit(X_dummies_train, y_label_train) 

print(f"Training Model Score: {classifier_logistic.score(X_dummies_train, y_label_train)}")
print(f"Testing Model Score: {classifier_logistic.score(X_dummies_test, y_label_test)}");


Training Model Score: 0.6485221674876848
Testing Model Score: 0.5253083794130158


C:\Users\rober\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [36]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
classifier_forest = RandomForestClassifier(random_state=42, n_estimators=200).fit(X_dummies_train, y_label_train)

print(f'Training Model Score: {classifier_forest.score(X_dummies_train, y_label_train)}')
print(f'Testing Model Score: {classifier_forest.score(X_dummies_test, y_label_test)}')

Training Model Score: 1.0
Testing Model Score: 0.623139089749043


Prediction Result: It appears my prediction was correct (at least for unscaled data), as both the Training Model and Testing Model Scores were higher for Random Forest. The results of the two models are below:

Logistical Regression
Training Model Score: 0.6485221674876848
Testing Model Score: 0.5253083794130158

Random Forest
Training Model Score: 1.0
Testing Model Score: 0.623139089749043

In [33]:
# Scale the data
scaler = StandardScaler().fit(X_dummies_train)
X_train_scaled = scaler.transform(X_dummies_train)
X_test_scaled = scaler.transform(X_dummies_test)

In [40]:
# Train the Logistic Regression model on the scaled data and print the model score
classifier_lr = LogisticRegression().fit(X_train_scaled, y_label_train)

print(f'Training Model Score: {classifier_lr.score(X_test_scaled, y_label_test)}')

Training Model Score: 0.7201190982560612


C:\Users\rober\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [39]:
# Train a Random Forest Classifier model on the scaled data and print the model score
classifier_rf = RandomForestClassifier(random_state=42, n_estimators=500).fit(X_train_scaled, y_label_train)

print(f'Testing  Model Score: {classifier_rf.score(X_test_scaled, y_label_test)}')

Testing  Model Score: 0.6150574223734581


Interestingly, the Logistical Regression Model is the more accurate model after scaling has been applied to the data. The LR model may be more effective at scaling and predicting a more linear set of data in an open set, while RF is better at predicting more variable data in an already established set of data.